datafusion/core/src/physical_optimizer/enforcement.rs

```diff
@@ -0,0 +1,2001 @@
+        let join_types = vec![
+            JoinType::Inner,
+            JoinType::Left,
+            JoinType::Right,
+            JoinType::Full,
+            JoinType::LeftSemi,
+            JoinType::LeftAnti,
+            JoinType::RightSemi,
+            JoinType::RightAnti,
+        ];
+
+        // Join on (a == b1)
+        let join_on = vec![(
+            Column::new_with_schema("a", &schema()).unwrap(),
+            Column::new_with_schema("b1", &right.schema()).unwrap(),
+        )];
+
+        for join_type in join_types {
+            let join = hash_join_exec(left.clone(), right.clone(), &join_on, &join_type);
+            let join_plan =
+                format!("HashJoinExec: mode=Partitioned, join_type={}, on=[(Column {{ name: \"a\", index: 0 }}, Column {{ name: \"b1\", index: 1 }})]", join_type);
+
+            match join_type {
+                JoinType::Inner
+                | JoinType::Left
+                | JoinType::Right
+                | JoinType::Full
+                | JoinType::LeftSemi
+                | JoinType::LeftAnti => {
+                    // Join on (a == c)
+                    let top_join_on = vec![(
+                        Column::new_with_schema("a", &join.schema()).unwrap(),
+                        Column::new_with_schema("c", &schema()).unwrap(),
+                    )];
+                    let top_join = hash_join_exec(
+                        join.clone(),
+                        parquet_exec(),
+                        &top_join_on,
+                        &join_type,
+                    );
+                    let top_join_plan =
+                        format!("HashJoinExec: mode=Partitioned, join_type={}, on=[(Column {{ name: \"a\", index: 0 }}, Column {{ name: \"c\", index: 2 }})]", join_type);
+
+                    let expected = match join_type {
+                        // Should include 3 RepartitionExecs
+                        JoinType::Inner | JoinType::Left | JoinType::LeftSemi | JoinType::LeftAnti => vec![
+                            top_join_plan.as_str(),
+                            join_plan.as_str(),
+                            "RepartitionExec: partitioning=Hash([Column { name: \"a\", index: 0 }], 10)",
+                            "ParquetExec: limit=None, partitions=[x], projection=[a, b, c, d, e]",
+                            "RepartitionExec: partitioning=Hash([Column { name: \"b1\", index: 1 }], 10)",
+                            "ProjectionExec: expr=[a@0 as a1, b@1 as b1, c@2 as c1, d@3 as d1, e@4 as e1]",
+                            "ParquetExec: limit=None, partitions=[x], projection=[a, b, c, d, e]",
+                            "RepartitionExec: partitioning=Hash([Column { name: \"c\", index: 2 }], 10)",
+                            "ParquetExec: limit=None, partitions=[x], projection=[a, b, c, d, e]",
+                        ],
+                        // Should include 4 RepartitionExecs
```

Comment:
why does a right outer join (and full join) need the extra repartition? Mostly I don't understand why you don't need the same repartition for the inner join as well

Reply:
This is a wonderful explanation -- thank you @mingmwang